src/caffe/layers/triplet_loss_layer.cpp

```diff
@@ -0,0 +1,191 @@
+  CHECK_EQ(bottom[5]->height(), 1);
+  CHECK_EQ(bottom[5]->width(), 1);
+  diff_pos.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  diff_neg.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  diff_par.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  diff_sq_pos.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  diff_sq_neg.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  diff_sq_par.Reshape(bottom[0]->num(), bottom[0]->channels(), 1, 1);
+  dist_sq_.Reshape(bottom[0]->num(), 1, 1, 1);
+  dist_sq_pos.Reshape(bottom[0]->num(), 1, 1, 1);
+  dist_sq_neg.Reshape(bottom[0]->num(), 1, 1, 1);
+  dist_sq_par.Reshape(bottom[0]->num(), 1, 1, 1);
+  // vector of ones used to sum along channels
+  summer_vec_.Reshape(bottom[0]->channels(), 1, 1, 1);
+  for (int i = 0; i < bottom[0]->channels(); ++i)
+    summer_vec_.mutable_cpu_data()[i] = Dtype(1);
+}
+
+template <typename Dtype>
+void TripletLossLayer<Dtype>::Forward_cpu(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top) {
+  int count = bottom[0]->count();
+  caffe_sub(
+      count,
+      bottom[0]->cpu_data(),  // a
+      bottom[1]->cpu_data(),  // b
+      diff_pos.mutable_cpu_data());  // a_i-b_i: positive
+  caffe_sub(
+      count,
+      bottom[0]->cpu_data(),  // a
+      bottom[2]->cpu_data(),  // c
+      diff_neg.mutable_cpu_data());  // a_i-c_i: negative
+  caffe_sub(
+      count,
+      bottom[3]->cpu_data(),  // d
+      bottom[4]->cpu_data(),  // e
+      diff_par.mutable_cpu_data());  // d_i-e_i: pair wise
+  const int channels = bottom[0]->channels();
+  Dtype margin = this->layer_param_.triplet_loss_param().margin();
+  Dtype loss(0.0);
+
+  for (int i = 0; i < bottom[0]->num(); ++i) {
+    // Triplet loss accumulation
+    // Loss component calculated from a and b
+    dist_sq_pos.mutable_cpu_data()[i] = caffe_cpu_dot(channels,
+        diff_pos.cpu_data() + (i*channels), diff_pos.cpu_data() + (i*channels));
+    // a b is a similar pair for triplet
+    dist_sq_.mutable_cpu_data()[i] = dist_sq_pos.cpu_data()[i];
+    // Loss component calculated from a and c
+    dist_sq_neg.mutable_cpu_data()[i] = caffe_cpu_dot(channels,
+        diff_neg.cpu_data() + (i*channels), diff_neg.cpu_data() + (i*channels));
+    // a c is a dissimilar pair for triplet
+    dist_sq_.mutable_cpu_data()[i] -= dist_sq_neg.cpu_data()[i];
+    // loss accumulated accumulated by the triplet part
+    loss += std::max(margin + dist_sq_.cpu_data()[i], Dtype(0.0));
+    // Pair wise loss accumulation
+    // Loss component calculated from d and e
+    dist_sq_par.mutable_cpu_data()[i] = caffe_cpu_dot(channels,
+        diff_par.cpu_data() + (i*channels), diff_par.cpu_data() + (i*channels));
+    // d e is a similar pair for pair wise
+    // loss accumulated by the pair wise part
+    loss += dist_sq_par.cpu_data()[i];
+  }
+  loss = loss / static_cast<Dtype>(bottom[0]->num()) / Dtype(2);
+  top[0]->mutable_cpu_data()[0] = loss;
+}
```

Comment:
I see you implemented the cost function as described in ["Learning Fine-grained Image Similarity with Deep Ranking"](http://arxiv.org/abs/1404.4661).
You could add a param to the loss layer to switch between this cost function and the one described in ["
Learning Descriptors for Object Recognition and 3D Pose Estimation"](http://arxiv.org/abs/1502.05908).
The function is similar but the authors report that theirs works better in special cases. So because the code could be very similar, could be worth implement also the newest cost  function